### ISSUE TITLE
New analysis suggestion - count of LV5, LV10, LV15 and LV20 patients 

### ISSUE THREAD

**Issue by vojtechhuser** [OHDSI/Achilles#36]
I have a new analysis. Is there a list of all SQL queries Achilles runs - I want to assign an ID to that new SQL

Is there a documentation on what output structure I should use for output of new analyses (new table ,existing table)

The analysis I want to add is this SQL query:  (in RedShift flavor)

(the query gets first visit and last visit for each patient, then rounds it to span years and arrives at span rounded to nearest lower multiple of 5 (substracts modulo results). It then counts number of patients in each bin.

``` sql
SET SEARCH_PATH TO ccae_cdm4;
select v5span years_span, count(*) from (
            select person_id
            --, fdt, ldt,(ldt-fdt) ,(ldt-fdt)/365.25 dif, cast((ldt-fdt)/365.25 as integer) ydif
            ,cast((ldt-fdt)/365.25 as integer)- mod(cast((ldt-fdt)/365.25 as integer),5) v5span
            from(
                    select person_id, min(e.visit_start_date) fdt, max(e.visit_end_date) ldt
                    from visit_occurrence e   
                    group by person_id
            )
)
 group by v5span order by v5span;
```

The output gives numbers like this:

```
0       99,110,955 
5       12,406,560 
10      676,189 
```

WHich is for CCAE and it means there are 99M patients with visit span of 0-5 years
12M patients with visit span 5-10 years and no patients with more than 15 years time span.

It is similar to existing analysis, but that is not a set of 4 numbers but a graph.

It is based on this work: 
https://github.com/OHDSI/sandbox/tree/master/CDMv4/analytical-SQL-code/vojtech_huser/IRIS

**Comment by chrisknoll** [comment]
Patrick, I was working with Vojetch on this offline, and one thing I realized is that the granularity of the existing analysis (108) which is 30d buckets doesn't translate well to the (365.25 * 5) day buckets that we'd want to do with the LV durations.  So, I'm suggesting that we add a new analysis that groups by 5 years.  If we instead want to create it into 1 year intervals, I can group the 1 year intervals into the 5 year intervals for the report that Vojtech is requesting. and if later we want to see different groupings (such as 1-3, 3-10, 10-15) we can slice up the by-year counts into whatever groups we want to plot.

Thoughts?

**Comment by pbr6cornell** [comment]
Wouldn't 30d*12mo/yr = 360d buckets adequately approximate the need?

**Comment by vojtechhuser** [comment]
I happy to re-use existing SQL queries. 

The problem I am having is to understand how many interim tables Achilles[Heel] is using. I have a gigantic SQL file  (Achilles.SQL) and no access to the achiles results DB or JSON.  (IMEDS lab)  

If someone would write one paragraph about how many "export tables" there are and their high level structure, it would help other people suggesting different  views of data from existing SQL queries.

Basically a little high level introduction to studying Achilles.SQL file.

**Comment by chrisknoll** [comment]
I was thinking about putting something substantial into the wiki as an introduction, but I can't dedicate the time I think it would take to do an adequate amount of information, so let me reply to you here with a short overview, then I'll think about how to capture the appropriate amount of detail for a Wiki page:

There are 4 tables in Achilles: 1 'meta' table which describes each analysis and strata names, and 3 tables that store the results tables that Achilles uses to do work.

ACHILLES_analysis (meta): This table stores analysis descriptions so it is easier to understand which analysis Id is calculating which datapoints.  

Example for Analysis 108 (count of pepole by length of obsrvation period in 30d increments:

```
insert into ACHILLES_analysis (analysis_id, analysis_name, stratum_1_name)
	values (108, 'Number of persons by length of observation period, in 30d increments', 'Observation period length 30d increments');
```

In this example, strata_1 is being decribed as 'observation period length in 30d increments' which his the only statistic captured in this analysis.

A different example where there is data with multiple levels of stratifcation:
```
insert into ACHILLES_analysis (analysis_id, analysis_name, stratum_1_name, stratum_2_name, stratum_3_name)
	values (116, 'Number of persons with at least one day of observation in each year by gender and age decile', 'calendar year', 'gender_concept_id', 'age decile');
```

This one counts people by year->gender->age decile.

The next 3 tables are the results tables:

ACHILLES_results: stores values that can be stratified by up to 5 levels.  strata_1 is always populated (because there's always on level to stratification), but some analysis have more (such as 116 described above).

ACHILLES_results_dist: this is similar to ACHILLES_results but specifically designed to store distribution values (min, max, q1, q3, median, stddev, avg), and also supports up to 5 levels of stratification so you see strat_1 -> strat_5 in the table.  

Note that the AHILLES_analysis does not specify if the analysis results goes into the _results or _results_dist table, you just need to look at the actual query to know.

ACHILLES_HEEL_results: just a simple table with an analysis ID and a text column for warnings.  This is more of a log table than anything that you could do additional analysis with.

So, to go back to your question/issue about understanding what 'export tables' there are, I would start by just looking at the list of analysis INSERT statements (if you do not have access to a database with Achilles directly) to understand what data-points we are capturing.  For this particular issue, we wanted to see 'number of people by 365*5 day intervals) but the closet one we can find by looking at the Analysis metadata is from 108: 'Number of persons by length of observation period, in 30d increments'.  So, if you don't have access to the IMEDs data, but you do have your own CDM instance you can copy the code from analysis 108 and execute it to get the results for analysis 108 into your own local environment:

```
select 108 as analysis_id,  floor(DATEDIFF(dd, op1.observation_period_start_date, op1.observation_period_end_date)/30) as stratum_1, COUNT_BIG(distinct p1.person_id) as count_value
from PERSON p1
	inner join 
	(select person_id, 
		OBSERVATION_PERIOD_START_DATE, 
		OBSERVATION_PERIOD_END_DATE, 
		ROW_NUMBER() over (PARTITION by person_id order by observation_period_start_date asc) as rn1
		 from OBSERVATION_PERIOD
	) op1
	on p1.PERSON_ID = op1.PERSON_ID
	where op1.rn1 = 1
group by floor(DATEDIFF(dd, op1.observation_period_start_date, op1.observation_period_end_date)/30)
;
```

Note this is in MS SQL server dialect so you might need to make changes to support your own platform.  Once you have this data, you can save it into a temp table on your side and decide how to aggregate that data into your 360d buckets (that's our only option if we are going to leverage the 30d intervals from this analysis).  That query looks like this:

```
select (stratum_1 / 12) as stratum_1,  sum(count_value) as count_value
from achilles_results
where analysis_id = 108
group by (stratum_1 / 12)  
```

But you wanted 360d * 5 (approx 5 years) intervals, no problem the query is actually:

```
select (stratum_1 / 60) as stratum_1,  sum(count_value) as count_value
from achilles_results
where analysis_id = 108
group by (stratum_1 / 60)
```

If you run this query on your own achilles_results instance (or create the achilles results by executing the analysis_id 108 query above), you'll get your intervals but at 1800d intervals (which is the 30d * 12m * 5y interval size).

Ok!  So that talks about the first part of your comment about 'what are the export tables'. we have the raw data, but how does it get into the website?  This is done in the second party of Achilles: exportToJson.

If you take a look at the exportToJson.R file in the R folder of the repository, you'll see that this file is 
the code that querys results from Achilles_results[_dist] and creates the JSON structures that we render on the website. For this example, we want to modify the 'Observation Period' reports so that function is found in generateObservationPeriodReport. Here's the place in the code where we use data from analysis 108:

```
  # 3.  Title: Length of observation
  # a.	Visualization:  bar
  # b.	Category:  length of observation period, 30d increments
  # c.	Values: # of persons

... some initialization code and sql query rendering...

  observationLengthStats <- querySql(conn,renderedSql)
...
  observationLengthHist$MIN = observationLengthStats$MIN_VALUE
  observationLengthHist$MAX = observationLengthStats$MAX_VALUE
  observationLengthHist$INTERVAL_SIZE = observationLengthStats$INTERVAL_SIZE
...  
  observationLengthData <- querySql(conn,renderedSql)
...
  observationLengthHist$DATA <- observationLengthData
```

The end of this code is a JSON node that contains a field 'observationLengthHist' which contains the MIN, MAX, INTERVAL_SIZE and INTERVALS. These are the datapoints required for plotting the data as a histogram. But in YOUR case, we would want to plot it as a piechart.  Let's take a look at one of the parts of code that does this:

```
  # 9.  Title:  Number of observation periods per person
  # a.	Visualization:  Pie
  # b.	Category:  Number of observation periods
  # c.	Values:  # of persons 
  
...
  personPeriodsData <- querySql(conn,renderedSql)
...
  output$PERSON_PERIODS_DATA = personPeriodsData
```

That is easy, but what's the actual query?
````
select row_number() over (order by ar1.stratum_1) as concept_id, 
	ar1.stratum_1 as concept_name, 
	ar1.count_value as count_value
from ACHILLES_results ar1
where ar1.analysis_id = 113
```

Aha! so, all we need to do is create a query that returns 3 columns: concept_id, concept_name, count_value.  Why concept_id/concept_name?  Because that's the fields that our pie chart library expects.  we can probably make that a bit more abstract, but live with it for now....

So, from your prior query, to get it into a form that can be used on the web:

```
select row_number() over (order by ar1.stratum_1) as conept_id, 
'LV' + cast(((ar1.stratum_1 + 1) * 5) as varchar(50)) as concept_name,
ar1.count_value as count_value
from (
  select (stratum_1 / 60) as stratum_1,  sum(count_value) as count_value
  from achilles_results
  where analysis_id = 108
  group by (stratum_1 / 60)
) ar1  
```

Results in:
```
conept_id	concept_name	count_value
1	LV5	3244
2	LV10	2158
3	LV15	594
```

(this is from our own internal sample set but it shows the example output).

If we wanted to get more fancy and limit it to say LV20+ we'd have to make the code a bit more robust, but this is a good start.

**Comment by t-abdul-basser** [comment]
Original question addressed by current documentation.
